### PR TITLE
Problem with unicode chars with add/modify contact

### PR DESCRIPTION
**Pull request by davec82**

### CODE REVIEW

File: ldaptor/delta.py [hunk 1]
```diff
@@ -8,7 +8,7 @@ changing of location in tree)
 from ldaptor import attributeset
 from ldaptor.protocols import pureldap, pureber
 from ldaptor.protocols.ldap import ldif, distinguishedname
-
+import codecs
```
Comment (antong) [inline ldaptor/delta.py:11]: Unused import? What is this for?

File: ldaptor/protocols/ldap/ldaperrors.py [hunk 2]
```diff
@@ -39,6 +39,11 @@ class Success(LDAPResult):
         pass
 
 class LDAPException(Exception, LDAPResult):
+
+    def _get_message(self): return self.__message
```
Comment (antong) [inline ldaptor/protocols/ldap/ldaperrors.py:43]: This I don't like. If I understand correctly the reason for these lines is to silence the warning:
    DeprecationWarning: BaseException.message has been deprecated as of Python 2.6

However, this property hack looks complicated and confusing and is just there to sweep the problem under the carpet. Silencing the warning does not make the 'message' attribute any less deprecated :-). The right way is to stop using 'message' and use 'args' instead I think.

File: ldaptor/protocols/ldap/ldapsyntax.py [hunk 3]
```diff
@@ -7,7 +7,7 @@ from ldaptor.protocols.ldap import ldapclient, ldif, distinguishedname, ldaperro
 from ldaptor.protocols import pureldap, pureber
 from ldaptor.samba import smbpassword
 from ldaptor import ldapfilter, interfaces, delta, attributeset, entry
-
+import codecs
```
Comment (antong) [inline ldaptor/protocols/ldap/ldapsyntax.py:10]: Another unused import?

**Comment by antong** [comment]
Thanks for contributing!

I see that your change encodes unicode strings to UTF-8 byte strings in two functions. I'm not sure I understand the use case. Can you give an example of what you are doing that requires or causes these routines get unicode strings?

Thanks,
Anton

File: ldaptor/delta.py (hunk 1 above)
Comment (davec82) [inline ldaptor/delta.py:11]: yes, unused import, sorry

File: ldaptor/protocols/ldap/ldapsyntax.py (hunk 3 above)
Comment (davec82) [inline ldaptor/protocols/ldap/ldapsyntax.py:10]: yes, unused import, sorry

**Comment by davec82** [comment]
Thanks for your reply.

When i call addChild or commit functions, attributes are passed in unicode format but when there is a char like è,ò, (a no ascii char), ldaptor can't add/modify contact because:

UnicodeEncodeError: 'ascii' codec can't encode character u'\xe8' in position 10: ordinal not in range(128)

You can try to set an attribute like  u'abcde\xe8' and see if error occurs.

Thanks

Davide

**Comment by antong** [comment]
Is it your code or ldaptor code that passes unicode strings to addChild()? Where does the unicode string come from? I think that instead of doing:
    myustring = u'èò'
    e['someattribute'] = myustring
    e.commit()

you should be doing:
    e['someattribute'] = myustring.encode('utf-8')
    e.commit()

I believe it may not be a good idea to automatically encode unicode strings to UTF-8. However, if you want to do it, I think it could be done more generally and not in arbitrary code paths as in your change. I pushed a completely untested change that does this to a new topic branch t/auto-utf8: antong/ldaptor@edc6d7d52edc79e82491. Does this do what you want?

Go ahead and test that change and say if you think it is a good idea.

Good luck and thanks!
Anton

**Pull request closed without merging**